### PR TITLE
[FLINK-2073] [ml] [docs] Adds contribution guide

### PR DESCRIPTION
**Pull request by tillrohrmann**
Adds contribution guide and implementation tutorial for new pipeline operators.

### CODE REVIEW

File: docs/libs/ml/contribution_guide.md [hunk 1]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+
+If you already know what you want to contribute to FlinkML all the better.
+It is still advisable to create a JIRA issue for your idea to tell the Flink community what you want to do, though.
+
+## Testing
+
+New contributions should come with tests to verify the correct behavior of the algorithm.
+The tests help to maintain the algorithm's correctness throughout code changes, e.g. refactorings.
+
+We distinguish between unit tests, which are executed during maven's test phase, and integration tests, which are executed during maven's verify phase.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Capitalization: maven -> Maven

File: docs/libs/ml/contribution_guide.md [hunk 2]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+    ...
+  }
+}
+{% endhighlight %}
+
+The test style does not have to be `FlatSpec` but can be any other scalatest `Suite` subclass. 
+
+## Documentation
+
+When contributing new algorithms, it is required to add code comments describing the functioning of the algorithm and its parameters with which the user can control its behavior.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: functioning -> functionality

File: docs/libs/ml/contribution_guide.md [hunk 3]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+Once you have implemented the algorithm with adequate test coverage and added documentation, you are ready to open a pull request.
+Details of how to open a pull request can be found [here](http://flink.apache.org/how-to-contribute.html#contributing-code--documentation). 
+
+## How to Implement a Pipeline Operator
+
+FlinkML follows the principle to make machine learning as easy and accessible as possible.
+Therefore, it supports a flexible pipelining mechanism which allows users to quickly define their analysis pipelines consisting of a multitude of different components.
+A pipeline operator is either a `Transformer` or a `Predictor`.
+A `Transformer` can be fitted to training data and transforms data from one format into another format.
+A scaler which changes the mean and variance of its input data according to the mean and variance of some training data is an example for a `Transformer`.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: example for -> example of

File: docs/libs/ml/contribution_guide.md [hunk 4]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+In contrast, a `Predictor` encapsulates a data model and the corresponding logic to train it.
+Once a `Predictor` has trained the model, it can be used to make new predictions.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Perhaps we should talk about transform() and predict(), and their different semantics. Then we can use the same line of thought when deciding what type of Estimator our intended algorithm is.

Does it make sense for a scaler to implement the predict function? How about a transform function? Then the choice between Transformer and Predictor becomes clear. OTOH it should be pretty self-explanatory which type of algorithm you are trying to implement, so we could avoid this for the sake of brevity.

File: docs/libs/ml/contribution_guide.md [hunk 5]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+object MeanTransformer {
+  case object Mean extends Parameter[Double] {
+    override val defaultValue: Option[Double] = Some(0.0)
+  }
+  
+  def apply(): MeanTransformer = new MeanTransformer
+}
+{% endhighlight %}
+
+Parameters are defined in the companion object of the transformer class and extend the `Parameter` class.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: But why are they case objects?

File: docs/libs/ml/contribution_guide.md [hunk 6]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+{% highlight scala %}
+class MeanTransformer extends Transformer[Centering] {}
+{% endhighlight %}
+
+Since we want to be able to configure the mean of the resulting data, we have to add a configuration parameter.
+
+{% highlight scala %}
+class MeanTransformer extends Transformer[Centering] {
+  def setMean(mean: Double): Mean = {
+    parameters.add(MeanTransformer.Mean, mu)
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: mu -> mean

File: docs/libs/ml/contribution_guide.md [hunk 7]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+
+{% highlight scala %}
+class MeanTransformer extends Transformer[Centering] {}
+{% endhighlight %}
+
+Since we want to be able to configure the mean of the resulting data, we have to add a configuration parameter.
+
+{% highlight scala %}
+class MeanTransformer extends Transformer[Centering] {
+  def setMean(mean: Double): Mean = {
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Mean -> this.type or MeanTransformer

File: docs/libs/ml/contribution_guide.md [hunk 8]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+    parameters.add(MeanTransformer.Mean, mu)
+  }
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Return this

File: docs/libs/ml/contribution_guide.md [hunk 9]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+The default value will be used if no other value has been set by the user of this component.
+If no default value has been specified, meaning that `defaultValue = None`, then the algorithm has to handle this situation accordingly.
+
+We can now instantiate a `MeanTransformer` object and set the mean value of the transformed data.
+But we still have to implement how the transformation works.
+The workflow can be separated into two phases.
+Within the first phase, the transformer learns the mean of the given training data.
+This knowledge can then be used in the second phase to transform the provided data with respect to the configured resulting mean value.
+
+The learning of the mean can be implemented within the `fit` operation of a `Transformer`.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Isn't fit technically an operation of Estimator?

File: docs/libs/ml/contribution_guide.md [hunk 10]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+        (left, right) => 
+          (left._1 + right._1, left._2 + right._2) 
+      }
+      .map{ p => (p._1/p._2).fromBreeze })
+  }
+}
+{% endhighlight %}
+
+If we look at the `transform` method in `Transformer`, we will see that we also need an implementation of `TransformOperation`.
+A possible mean transforming implementation could look the following.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: look -> look like

File: docs/libs/ml/contribution_guide.md [hunk 11]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+A general description of FlinkML's pipelining can be found [here]({{site.baseurl}}/libs/ml/pipelines.html).
+In order to support the pipelining, algorithms have to adhere to a certain design pattern, which we will describe next.
+
+Let's assume that we want to implement a pipeline operator which changes the mean of your data.
+At first, we have to reflect which type of pipeline operator it is.
+Since centering data is a common preprocessing step in any analysis pipeline, we will implement it as a `Transformer`.
+Therefore, we first create a `MeanTransformer` class which inherits from `Transformer`
+
+{% highlight scala %}
+class MeanTransformer extends Transformer[Centering] {}
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Transformer[Centering] -> Transformer[MeanTransformer]

File: docs/libs/ml/contribution_guide.md [hunk 12]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+{% endhighlight %}
+
+Since we want to be able to configure the mean of the resulting data, we have to add a configuration parameter.
+
+{% highlight scala %}
+class MeanTransformer extends Transformer[Centering] {
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Transformer[Centering] -> Transformer[MeanTransformer]

File: docs/libs/ml/contribution_guide.md [hunk 13]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+
+{% highlight scala %}
+object MeanTransformer {
+  implicit val labeledVectorFitOperation = new FitOperation[MeanTransformer, LabeledVector] ...
+  
+  implicit val labeledVectorTransformOperation = new TransformOperation[MeanTransformer, LabeledVector, LabeledVector] ...
+}
+{% endhighlight %}
+
+If we wanted to implement a `Predictor` instead of a `Transformer`, then we would have to do the same.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: to do the same -> to work in a similar way.

File: docs/libs/ml/contribution_guide.md [hunk 14]
```diff
@@ -20,7 +21,329 @@ specific language governing permissions and limitations
...
+
+{% highlight scala %}
+val trainingData: DataSet[DenseVector] = ...
+val meanTransformer = MeanTransformer()
+
+meanTransformer.fit(trainingData)
+{% endhighlight %}
+
+we receive the following error at runtime: `"There is no FitOperation defined for class MeanTransformer which trains on a DataSet[org.apache.flink.ml.math.DenseVector]"`.
+The reason is that the Scala compiler could not find a fitting `FitOperation` value with the right type parameters for the implicit parameter of the `fit` method.
```
Comment (thvasilo) [inline docs/libs/ml/contribution_guide.md]: Is it the compiler that detects this?

**Comment by thvasilo** [comment]
Looks good, just minor comments, plus the HTML code on the title fix.

File: docs/libs/ml/contribution_guide.md (hunk 1 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: Good catch.

File: docs/libs/ml/contribution_guide.md (hunk 2 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: That's German ;-)

File: docs/libs/ml/contribution_guide.md (hunk 3 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: :+1:

File: docs/libs/ml/contribution_guide.md (hunk 8 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: :+1:

File: docs/libs/ml/contribution_guide.md (hunk 9 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: Yes you're right.

File: docs/libs/ml/contribution_guide.md (hunk 10 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: :+1:

File: docs/libs/ml/contribution_guide.md (hunk 14 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: Yes the compiler looks for implicit values in order to give them to implicit parameters. If it does not find one, then it takes our fallback implicit value.

File: docs/libs/ml/contribution_guide.md (hunk 13 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: agreed

File: docs/libs/ml/contribution_guide.md (hunk 12 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: changed

File: docs/libs/ml/contribution_guide.md (hunk 11 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: good catch

File: docs/libs/ml/contribution_guide.md (hunk 7 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: `this.type` is good

File: docs/libs/ml/contribution_guide.md (hunk 6 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: changed

File: docs/libs/ml/contribution_guide.md (hunk 5 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: That is true. I'll explain it.

File: docs/libs/ml/contribution_guide.md (hunk 4 above)
Comment (tillrohrmann) [inline docs/libs/ml/contribution_guide.md]: I made it more explicit. At least I hope.

**Comment by tillrohrmann** [comment]
I addressed your comments.

**Comment by thvasilo** [comment]
LGTM, only one minor comment. I also like the addition of descriptions for the fit and predict methods to each algorithm.

**Comment by tillrohrmann** [comment]
Will merge it then.

**Pull request closed without merging**